File: ServerSide/BallPassSchedulePattern/Players_QrcodeReaderTypes/Decode_d_RightBottomReaderPlayerDir/RightBottomReaderPlayer.py
import os, sys
sys.path.append("../..")
from Players_CommonPlayers.SuperPlayerDir.SuperPlayer import SuperPlayer

class RightBottomReaderPlayer(SuperPlayer):  # 名前はりなに決めてもらってもOKよ！
    def __init__(self):
        super().__init__()  # スーパークラスの初期化メソッドを呼び出す
        self.my_name = None  # プレイヤーの名前
        self.data_read = []  # データを読み込んで保持する変数
        self.replaced_matrix = None # 置き換え後のmatrix
        self.png_file_path = None # 写真のパス
        self.binary_matrix_2Dlist = None # 置き換え前のmatrix
        self.new_list2d = None # データ部分をn*2のリストに変換したもの


    def return_my_name(self):
        return "RightBottomReaderPlayer"  # ここも任意の名前で変更可能よ
    
    def flatten_list2d(self, list2d=None):
        """ 2次元リストを1次元リストにするメソッド
        ・shape的に(n, 2)の2次元リストに限る
        """
        
        """ 入力 """
        list2d
        
        """ 初期化 """
        read_list = [] # 読み込まれた1次元リスト(最終的に完成品となる)
        
        """ メイン処理 """
        # 全体を回す。
        for pair in list2d: # 1行取る
            
            """ 右, 左の順に格納。負の数である場合は格納しないでパスする。 """
            # 右から見る。
            if 0 <= pair[1]: # 0以上であれば
                read_list.append(pair[1]) # 登録
            else:
                pass
                
            # 次に左を見る。
            if 0 <= pair[0]: # 0以上であれば
                read_list.append(pair[0]) # 登録
            else:
                pass
        
        """ 出力 """
        return read_list
    

    def main(self):
        """
        このメソッド実行直前に、
        このスーパークラスのメンバ変数 one_time_world_instanceに、
        最新のworldインスタンスを代入しています。
        """
        
        """ 入力 """
        woP = self.one_time_world_instance.columnSplitterConcatPlayer
        self.replaced_matrix = woP.replaced_matrix # 置き換え後のmatrix
        self.png_file_path = woP.png_file_path # 写真のパス
        self.binary_matrix_2Dlist = woP.binary_matrix_2Dlist # 置き換え前のmatrix
        self.new_list2d = woP.new_list2d # データ部分をn*2のリストに変換したもの

        
        
        # n行2列のマトリックスを取得
        binary_matrix = self.binary_matrix_2Dlist
        
        """ メイン処理 """
        # データを右下から読み込む
        self.data_read = self.flatten_list2d(list2d=binary_matrix)
        # 読み込んだデータを確認
        print(f"データが右下から読み込まれました: {self.data_read}")

        """ 出力 """
        self.replaced_matrix # 置き換え後のmatrix
        self.png_file_path # 写真のパス
        self.new_list2d # データ部分をn*2のリストに変換したもの
        self.data_read # データ部分を1次元リスト化したもの。

        # 自身を更新
        self.one_time_world_instance.rightBottomReaderPlayer = self

        return "Completed"

File: InstanceSide/BallPassSchedulePattern/Players_QrcodeReaderTypes/Decode_d_RightBottomReaderPlayerDir/RightBottomReaderPlayer.cs
using UdonSharp;
using UnityEngine;

public class RightBottomReaderPlayer : SuperPlayer
{
    public int[][] replacedMatrix; // 置き換え後のマトリックス
    public string pngFilePath; // 写真のパス
    public int[][] binaryMatrix2DList; // 置き換え前のマトリックス
    public int[][] newList2D; // n*2のリストに変換したもの
    public int[] dataRead; // 右下から読み込んだデータを保持する1次元リスト

    public RinaNumpy rinaNumpy; // RinaNumpyインスタンスをアタッチ

    void Start()
    {
        ResetPlayer(); // プレイヤーの初期化
    }

    public void ResetPlayer()
    {
        myName = "RightBottomReaderPlayer"; // プレイヤー名を設定
        replacedMatrix = new int[0][]; // 初期化
        pngFilePath = "";
        binaryMatrix2DList = new int[0][]; // 初期化
        newList2D = new int[0][]; // 初期化
        dataRead = new int[0]; // 初期化
    }

    public override string ReturnMyName()
    {
        return "RightBottomReaderPlayer";
    }

    private int[] FlattenList2D(int[][] list2D)
    {
        // 2次元リストを1次元リストに変換する
        int[] flattenedList = new int[0]; // 最終的な結果
        int totalLength = list2D.Length * 2; // 合計の長さを計算

        int index = 0;
        int[] tempFlattenedList = new int[totalLength]; // 一時的な配列を用意

        for (int i = 0; i < list2D.Length; i++)
        {
            int[] pair = list2D[i];

            // 右からデータをチェックして格納
            if (pair[1] >= 0)
            {
                tempFlattenedList[index++] = pair[1];
            }

            // 次に左をチェックして格納
            if (pair[0] >= 0)
            {
                tempFlattenedList[index++] = pair[0];
            }
        }

        // 必要な長さに切り詰めて新しい配列を作成
        flattenedList = rinaNumpy.SubArray(tempFlattenedList, 0, index); // RinaNumpyのSubArrayを使用

        return flattenedList;
    }

    public override string ExecuteMain()
    {
        // ColumnSplitterConcatPlayerの結果を取得
        ColumnSplitterConcatPlayer woP = (ColumnSplitterConcatPlayer)oneTimeWorldInstance.GetComponent(typeof(ColumnSplitterConcatPlayer));

        if (woP == null)
        {
            Debug.LogError("ColumnSplitterConcatPlayer is not attached or initialized.");
            return "Error";
        }

        // データを取得
        replacedMatrix = woP.replacedMatrix;
        pngFilePath = woP.pngFilePath;
        binaryMatrix2DList = woP.binaryMatrix2DList;
        newList2D = woP.newList2D;

        // 右下からデータを読み取る
        dataRead = FlattenList2D(binaryMatrix2DList);

        // 読み込んだデータをログで確認
        Debug.Log("データが右下から読み込まれました: " + rinaNumpy.IntArrayToString(dataRead)); // RinaNumpyのIntArrayToStringを使用

        // 自身を更新
        oneTimeWorldInstance.rightBottomReaderPlayer = this;

        return "Completed";
    }
}
